### DISCUSSION TITLE
How to make theorem-like environments sharing the same counter

### DISCUSSION THREAD

**Discussion by finkelshtein** [quarto-dev/quarto-cli#10844] [Q&A]
### Description

Hi,
I have several lecture notes prepared in Latex, and I would like to transform them into a Quarto book with PDF and HTML outputs.
One of my main issues is the following: I heavily use the standard feature of **amsthm** package of sharing counters for theorem-like environments:
``` latex
\newtheorem{theorem}{Theorem}[section]
\newtheorem{definition}[theorem]{Definition}
\newtheorem{remark}[theorem]{Remark}
```
would make something like Theorem 1, Definition 2, Theorem 3, Remark 4, etc.

My aim is to reproduce the attached picture
<img src="https://github.com/user-attachments/assets/da6ee2c4-66bf-4fe9-b449-1bbdcf542dbd" width=50% height=50%>

As you can see, here there are several "unusual" features:

1. Shared counters for theorem-like environment.
2. Creating a new theorem-like environment (I called it "note") which does not have a prefix (like "Theorem" or "Definition") but still (may) have a title and shares the counter.
3. The theorem-like titles are not in parenthesis and are also separated from the theorem's text.
4. (Differently) coloured vertical bars outside the theorems.

In Latex, this can all be done using **tcolorbox** package. In Quarto, I'd prefer to have just:

````

::: {#def-first}

## Important definition

A number $a$ is called *positivie* if $a>0$.

::: 

::: {#thm-first}

## Important theorem

If $a>b$ and $b>c$ then $a>c$.

::: 

::: {#rem-first}

## Important remark

The property in @thm-first is called ``transitivity''.

::: 
    
::: {#nt-first}

## Commutation relations

We are going to discuss now commutation relations.

::: 

````

and then for the PDF output the rest would be done in latex preamble. For HTML, I understand that item 4 can be done using CSS file. However, for 1 and probably 3, I would need some lua filter, I guess.

Hence, the first question is indeed, how to make several theorem-like environments sharing the same counter?

**Comment by cscheid** [comment]
You'll need a fair bit of LaTeX and Quarto expertise to do this, unfortunately.

In HTML formats, Quarto does all of the crossref counter work; you'll need to reimplement that. On top of it, you want this to work on _books_, which are a particularly complicated part of the codebase, since global state needs to be maintained across different Pandoc renders.

In LaTeX formats, Quarto defers the crossref work to LaTeX packages. In order to share counters across these different environments, you'll probably have to write your own environments.

To do this in a single page for HTML and LaTeX formats, it will take a combination of filter work and LaTeX packages. Unfortunately, I'm not sure it's possible to make this work in a _book_ at the moment.

**Reply by finkelshtein** [reply]
Thanks! 
Latex part is familiar for me (and the picture above is a simplified example of my work of course, where a lot of sophisticated environments were written). 
However, I know very little about how Quarto operates with crossref work. Does exist some kind of documentation/workflow description/notes about how Quarto operates, how lua-filters are applied, when Pandoc acts etc?

**Reply by cscheid** [reply]
> However, I know very little about how Quarto operates with crossref work. Does exist some kind of documentation/workflow description/notes about how Quarto operates, how lua-filters are applied, when Pandoc acts etc?

Unfortunately, not yet. We're working on that right now but it's not really ready for public consumption.

If you want to go down that route you're going to have to download the dev version of Quarto and learn how our filter chain works, in [`src/resources/filters/main.lua`](https://github.com/quarto-dev/quarto-cli/tree/main/src/resources/filters/main.lua) etc.
